### ISSUE TITLE
Internal Error Connecting - CockroachDB

### ISSUE THREAD

**Issue by malexdev** [dbeaver/dbeaver#5067]
#### System information: 
- Operating system (distribution) and version: Windows 7
- DBeaver version: 5.3.3 Community
- Java version: 1.8.0_201 x86_64
- Additional extensions: None that I know of

#### Connection specification:
- Database name and version: CockroachDB 2.1.3
- Driver name: CockroachDB
- Do you use tunnels or proxies (SSH, SOCKS, etc)? No

#### Describe the problem you're observing:
I configured the connection to CockroachDB via the steps provided here: https://www.cockroachlabs.com/docs/v2.1/third-party-database-tools.html

This fails to connect with the text `Internal error connecting to jdbc:postgresql://<host>:26257/system`

I have done nothing more than install DBeaver, then immediately attempt to follow those steps - no other configuration changes to DBeaver.

#### Steps to reproduce, if exist:

1. Configure CockroachDB as a secure instance.
1. Install DBeaver fresh.
1. Add a new CockroachDB connection as specified in https://www.cockroachlabs.com/docs/v2.1/third-party-database-tools.html
1. Note that the connection fails due to an internal error.

I also must note that the root certificate is a self-signed certificate that is installed in the Windows certificate store and was added to the system Java control panel.

#### Include any warning/errors/backtraces from the logs

From `dbeaver-debug.log`: 
```
2019-01-21 13:55:46.996 - Connect with 'jdbc:postgresql://<host>:26257/system' (postgres-cockroach-jdbc-16871fbb1a8-26b2ce0b22a5d680)
2019-01-21 13:55:47.057 - Connection failed (postgres-cockroach-jdbc-16871fbb1a8-26b2ce0b22a5d680)
2019-01-21 13:55:47.060 - org.jkiss.dbeaver.DBException: Internal error connecting to jdbc:postgresql://<host>:26257/system
org.jkiss.dbeaver.DBException: Internal error connecting to jdbc:postgresql://<host>:26257/system
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:826)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:70)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionWizard$ConnectionTester.run(ConnectionWizard.java:232)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:101)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.NullPointerException
	at org.jkiss.dbeaver.ext.postgresql.model.PostgreDataSource.initializeRemoteInstance(PostgreDataSource.java:149)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:100)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:89)
	at org.jkiss.dbeaver.ext.postgresql.model.PostgreDataSource.<init>(PostgreDataSource.java:80)
	at org.jkiss.dbeaver.ext.postgresql.PostgreDataSourceProvider.openDataSource(PostgreDataSourceProvider.java:97)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:770)
	... 4 more
2019-01-21 13:59:00.132 - > Start org.eclipse.ui.views.log [1.2.300.v20180827-1331]
2019-01-21 13:59:11.847 - Connect with 'CockroachDB - postgres' (postgres-cockroach-jdbc-16871e5e3eb-450b6b9de3e7ea56)
2019-01-21 13:59:11.893 - Connection failed (postgres-cockroach-jdbc-16871e5e3eb-450b6b9de3e7ea56)
2019-01-21 13:59:11.896 - org.jkiss.dbeaver.DBException: Internal error connecting to CockroachDB - postgres
org.jkiss.dbeaver.DBException: Internal error connecting to CockroachDB - postgres
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:826)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:70)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.runSync(ConnectJob.java:98)
	at org.jkiss.dbeaver.ui.actions.datasource.DataSourceHandler.connectToDataSource(DataSourceHandler.java:107)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.initConnection(DataSourceDescriptor.java:665)
	at org.jkiss.dbeaver.model.navigator.DBNDataSource.initializeNode(DBNDataSource.java:147)
	at org.jkiss.dbeaver.model.navigator.DBNDatabaseNode.getChildren(DBNDatabaseNode.java:204)
	at org.jkiss.dbeaver.model.navigator.DBNDatabaseNode.getChildren(DBNDatabaseNode.java:1)
	at org.jkiss.dbeaver.ui.navigator.NavigatorUtils.getNodeChildrenFiltered(NavigatorUtils.java:560)
	at org.jkiss.dbeaver.ui.navigator.database.load.TreeLoadService.evaluate(TreeLoadService.java:49)
	at org.jkiss.dbeaver.ui.navigator.database.load.TreeLoadService.evaluate(TreeLoadService.java:1)
	at org.jkiss.dbeaver.ui.LoadingJob.run(LoadingJob.java:86)
	at org.jkiss.dbeaver.ui.LoadingJob.run(LoadingJob.java:71)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:101)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.lang.NullPointerException
	at org.jkiss.dbeaver.ext.postgresql.model.PostgreDataSource.initializeRemoteInstance(PostgreDataSource.java:149)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:100)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:89)
	at org.jkiss.dbeaver.ext.postgresql.model.PostgreDataSource.<init>(PostgreDataSource.java:80)
	at org.jkiss.dbeaver.ext.postgresql.PostgreDataSourceProvider.openDataSource(PostgreDataSourceProvider.java:97)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:770)
	... 14 more
```

**Comment by serge-rider** [comment]
I can't reproduce this with my local CockroachDB instance. I believe we'll cooperate with CDB team to reproduce and resolve this.

**Comment by malexdev** [comment]
@serge-rider Please let me know if I can do anything to assist.
Just now I also tested this using the MacOS version of DBeaver, with the same results, if that helps.

Thanks very much for your time!

**Comment by knz** [comment]
@malexdev I too have checked this, using dbeaver 5.3 Enterprise and CockroachDB 2.1.3. The error does not occur.

Can you provide more details:

1) on which platform are you running CockroachDB; note that Windows is not supported to run the db server at this time
2) is dbeaver actually trying to connect to CockroachDB at all? Please check the server log files for anything suspicious
3) is it possible that you are running another service on the same port number? it's possible that you are running a new version of cockroachdb somewhere, while running an old version somewhere else (and attempting to connect to the latter)

Thanks

**Comment by knz** [comment]
Oh I think that's the difference with our end -- I haven't tested with TLS certificates loaded, and perhaps Serge hasn't either.

I'll try again with certs, or maybe @serge-rider you can too?

**Comment by malexdev** [comment]
I tried with an insecure cluster and was able to log in.
However, of course we can't use an insecure cluster in production.

At this point I'm wondering if my self signed CA certificate is not supported by DBeaver.
I'll try having CRDB generate the CA cert.

**Comment by serge-rider** [comment]
I believe this is related to Java-specific certificate validation. This is definitely not DBeaver requirement but PostgreSQL JDBC driver requirement.

@malexdev Coudl you try to set factory to `org.postgresql.ssl.NonValidatingFactory` with self signed certificate?

**Comment by serge-rider** [comment]
Reopening as we don't have any good solution for this yet

**Comment by malexdev** [comment]
Thanks for the additional investigation, @serge-rider.

I reapplied my custom CA cert and everything worked flawlessly.
That's when I realized that when I redid my configuration for my test earlier, I was using the `root` user instead of the user I was initially testing with.
From there, I drilled down and discovered that I had never given the user permission to access any databases, so I am now assuming that "this user has no access to anything" was the "internal error" that DBeaver encountered.

This issue didn't appear when using `cockroach sql` because it let me connect with that user.  Based on that I didn't catch the permissions issue, but I assume that if I had actually tried to run any queries they would have failed and I would have seen the problem much faster.

Thanks so much for your time and effort. As of now, everything seems to be working exactly as I'd like it to work, and the issue was my own faulty configuration of the system.
